component/kafka/group/component.go

```diff
@@ -0,0 +1,425 @@
+// The default failure strategy is the ExitStrategy.
+// The default batch size is 1 and the batch timeout is 100ms.
+// The default number of retries is 0 and the retry wait is 0.
+func New(name, group string, brokers, topics []string, proc kafka.BatchProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	var errs []error
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if group == "" {
+		errs = append(errs, errors.New("consumer group is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if validation.IsStringSliceEmpty(topics) {
+		errs = append(errs, errors.New("topics are empty or have an empty value"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	defaultSaramaCfg, err := kafka.DefaultSaramaConfig(name)
+	if err != nil {
+		return nil, err
+	}
+
+	cmp := &Component{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: defaultSaramaCfg,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         kafka.BatchProcessorFunc
+	failStrategy kafka.FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+	commitSync   bool
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	return c.processing(ctx)
+}
+
+func (c *Component) processing(ctx context.Context) error {
+	var componentError error
+
+	retries := int(c.retries)
+	for i := 0; i <= retries; i++ {
+		handler := newConsumerHandler(ctx, c.name, c.group, c.proc, c.failStrategy, c.batchSize,
+			c.batchTimeout, c.commitSync)
+
+		client, err := sarama.NewConsumerGroup(c.brokers, c.group, c.saramaConfig)
+		if err != nil {
+			componentError = err
+			log.Errorf("error creating consumer group client for kafka component: %v", err)
+		}
+
+		if client != nil {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, c.topics, handler); err != nil {
+				log.Errorf("error from kafka consumer: %v", err)
```

Comment:
This is the same `err` set on `handler.err`, correct?

Reply:
Nope, it can be a connection error if the handler never processed anything for example. It can be a rebalance error, whatever.

Reply:
Oh, okay. Wouldn't this be a better error to return when retries are exhausted then?

Reply:
I am thinking what if we return `nil` there because of this failure here

Reply:
the `handler.err` is never exposed here actually. Tested that scenario. The error here doesn't happen even if Kafka disconnects somehow. I will capture the err here in `consumerError` actually since it might be useful.

Reply:
Thanks; I expect it will be useful during troubleshooting to clear out some doubts e.g. "it's not that, because the consumer error is `x`"

Reply:
I fixed some cases of assigning `consumerError`.. now it seems well enough